File: storm/bolts/anomaly_bolt.py
"""
AnomalyBolt skeleton: compare counts against threshold and emit anomalies.
"""
from streamparse.bolt import Bolt
import yaml


class AnomalyBolt(Bolt):
    def initialize(self, conf, ctx):
        cfg_path = conf.get('config_path', 'config/config.yaml')
        with open(cfg_path, 'r') as f:
            cfg = yaml.safe_load(f)
        self.threshold = int(cfg.get('storm', {}).get('anomaly_threshold', 25))
        sev = cfg.get('storm', {}).get('severity', {})
        self.low = int(sev.get('low_min', 25))
        self.medium = int(sev.get('medium_min', 40))
        self.high = int(sev.get('high_min', 60))

    def process(self, tup):
        district, count, window_start, window_end = tup.values
        if count < self.low:
            return

        if count >= self.high:
            severity = 'HIGH'
        elif count >= self.medium:
            severity = 'MEDIUM'
        else:
            severity = 'LOW'

        anomaly = {
            'district': str(district),
            'event_count': int(count),
            'window_start': window_start,
            'window_end': window_end,
            'severity': severity
        }
        self.emit([anomaly])

File: java-topology/src/main/java/com/project/bolts/AnomalyBolt.java
package com.project.bolts;

import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.topology.base.BaseRichBolt;
import org.apache.storm.task.TopologyContext;
import org.apache.storm.tuple.Fields;
import org.apache.storm.tuple.Tuple;
import org.apache.storm.tuple.Values;

import java.util.HashMap;
import java.util.Map;

public class AnomalyBolt extends BaseRichBolt {
    private final int threshold;
    private final int lowMin;
    private final int mediumMin;
    private final int highMin;
    private org.apache.storm.task.OutputCollector collector;

    public AnomalyBolt(int threshold, int lowMin, int mediumMin, int highMin) {
        this.threshold = threshold;
        this.lowMin = lowMin;
        this.mediumMin = mediumMin;
        this.highMin = highMin;
    }

    @Override
    public void prepare(java.util.Map stormConf, TopologyContext context, org.apache.storm.task.OutputCollector collector) {
        this.collector = collector;
    }

    @Override
    public void execute(Tuple tuple) {
        String district = tuple.getStringByField("district");
        int count = tuple.getIntegerByField("count");
        long windowStart = tuple.getLongByField("window_start");
        long windowEnd = tuple.getLongByField("window_end");

        String severity = null;
        if (count >= highMin) {
            severity = "HIGH";
        } else if (count >= mediumMin) {
            severity = "MEDIUM";
        } else if (count >= lowMin) {
            severity = "LOW";
        }

        Map<String, Object> anomaly = new HashMap<String, Object>();
        anomaly.put("district", district);
        anomaly.put("event_count", count);
        anomaly.put("window_start", windowStart);
        anomaly.put("window_end", windowEnd);
        anomaly.put("severity", severity);
        anomaly.put("threshold_value", threshold);
        collector.emit(new Values(anomaly));
        collector.ack(tuple);
    }

    @Override
    public void declareOutputFields(OutputFieldsDeclarer declarer) {
        declarer.declare(new Fields("anomaly"));
    }
}